### ISSUE TITLE
Lambda function access to outer environment

### ISSUE THREAD

**Issue by SebastianS90** [Whiley/WhileyCompiler#614]
So this works totally as expected if you know how to implement anonymous methods in a compiler:

```
type fun is function()->(int)

method main(whiley.lang.System.Console console):
    int x = 1
    fun f = &(->x)
    console.out.println(f())
    x = 2
    console.out.println(f())
```

Prints `1 1`

But it might not be what a normal user expects to happen.
And indeed, Java makes a [nice workaround](https://docs.oracle.com/javase/specs/jls/se7/html/jls-8.html#jls-8.1.3) for that:

> Any local variable, formal parameter, or exception parameter used but not declared in an inner class must be declared final.

So they forbid the second assignment to x.

Maybe it's more a specification issue (explicitly state that on creation of the lambda all values get copied). 
But I thought it was worth mentioning.

**Comment by DavePearce** [comment]
Hmmm, maybe introducing <code>final</code> makes sense for this.  Dunno.

**Comment by DavePearce** [comment]
This is fixed now, right?

**Comment by SebastianS90** [comment]
@DavePearce Not really, what I fixed is #612 (that you cannot use the lambda parameters later due to a type checker bug).

But here we need `final` for a clean solution as in Java.

Additional to the program above we get confusion when we introduce `x++` expressions:
````
int x = 1
mylambda fun = &(-> x++)
int y = fun()
assert x == 1
assert y == 2
````

OK, we don't have `x++` yet and as lambdas cannot contain statements, therefore you cannot modify the outside `x` within a lambda. But in case you can, I strongly guess it will modify a copy and not affect the outer `x`, which might be surprising to the programmer who expects `x == 2` after the program above.

Copy-behavior of values accessed by lambdas should either be clearly documented or made clear by using `final` (then it doesn't matter if you use a copied value or the original one - it cannot change as it is `final`)

**Comment by DavePearce** [comment]
Yup, I won't be including <code>x++</code> for a few reasons, including this :)

Hmmm, not sure what I really think about this.  Personally I think it's OK as is.  But, let's leave for now to think about ...
